### PR TITLE
Kick-off of API Blueprint module for apigility

### PR DESCRIPTION
**Pull request by miiila**

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+machine:
+  php:
+    version: 5.3.25
```
Comment (weierophinney) [inline circle.yml]: What is this file for? (I'm assuming it's for a CI service of some sort, but I don't know which one).

Also, if this is for a CI service, we typically test from our minimum supported version (currently 5.3, but we'll be bumping to 5.5 soon) through most stable, plus HHVM and PHP 7 nightlies; is the service this is for capable of testing multiple versions? If so, please set that up.

Comment (weierophinney) [inline circle.yml]: Also, if it *is* for a CI service, let me know so I can setup the integrations with this repo. :)

File: composer.json [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            "dev-master": "1.0-dev",
+            "dev-develop": "1.1-dev"
+        }
+    },
+    "require": {
+        "php": ">=5.3.23",
+        "zendframework/zend-eventmanager": "~2.3",
+        "zendframework/zend-mvc": "~2.3",
+        "zendframework/zend-servicemanager": "~2.3",
+        "zendframework/zend-view": "~2.3",
```
Comment (weierophinney) [inline composer.json]: We should specify `~2.4` at this point (we made that switch with v1.1).

File: composer.json [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        "zfcampus/zf-apigility-documentation": "~1.0",
+        "zfcampus/zf-content-negotiation": "~1.0"
+    },
+    "require-dev": {
+        "phpunit/PHPUnit": "3.7.*"
+    },
+    "autoload": {
+        "psr-4": {
+            "ZF\\Apigility\\Documentation\\ApiBlueprint\\": "src/",
+            "ZFTest\\Apigility\\Documentation\\ApiBlueprint\\": "test/"
```
Comment (weierophinney) [inline composer.json]: Move this one to an `autoload-dev` section.

File: config/module.config.php [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            ),
+        ),
+    ),
+
+    'service_manager' => array(
+        'factories' => array(
+            'ZF\Apigility\Documentation\ApiBlueprint\ApiBlueprintViewStrategy' => 'ZF\Apigility\Documentation\ApiBlueprint\ApiBlueprintViewStrategyFactory',
+        ),
+        'services' => array(
+            'ZF\Apigility\Documentation\ApiBlueprint\ApiBlueprintViewRenderer' => new ZF\Apigility\Documentation\ApiBlueprint\ApiBlueprintRenderer(),
```
Comment (weierophinney) [inline config/module.config.php]: In order to allow caching of configuration, this should be done as an invokable instead.

File: config/module.config.php [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+            'ZF\Apigility\Documentation\ApiBlueprint\Controller' => 'ZF\Apigility\Documentation\ApiBlueprint\ControllerFactory',
+        ),
+    ),
+    'zf-content-negotiation' => array(
+        'controllers' => array(
+            'ZF\Apigility\Documentation\ApiBlueprint\Controller' => 'Documentation',
+        ),
+        'accept_whitelist' => array(
+            'ZF\Apigility\Documentation\ApiBlueprint\Controller' => array(
+                0 => 'text/vnd.apiblueprint+markdown',
```
Comment (weierophinney) [inline config/module.config.php]: You can remove the index here.

File: phpunit.xml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<phpunit
+	bootstrap="./test/Bootstrap.php">
```
Comment (weierophinney) [inline phpunit.xml]: Let's make this simpler: have it load "./vendor/autoload.php", and then remove `test/Bootstrap.php`. All the bootstrap is doing is setting up autoloading, so we really don't need it. (I'm going to be updating the various other modules to do this as well; it's a remnant from old ZF2 practices, which we recently removed with the component split.)

File: src/Action.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
+<?php 
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (weierophinney) [inline src/Action.php]: 2015 instead of 2014.

File: src/Action.php [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+ * @copyright Copyright (c) 2015 Apiary Ltd. <support@apiary.io>
+ */
+
+namespace ZF\Apigility\Documentation\ApiBlueprint;
+
+use ZF\Apigility\Documentation\Operation as BaseOperation;
+
+class Action {
```
Comment (weierophinney) [inline src/Action.php]: We follow [PSR-1](http://www.php-fig.org/psr/psr-1/) and [PSR-2](http://www.php-fig.org/psr/psr-2/); opening brace goes on the following line. I can fix these sorts of CS issues when I merge, though. (I'll also add a dependency on php-codesniffer when I do, which will provide the tools for detecting and fixing these.)

File: src/Action.php [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+	/**
+	 * @var BaseOperation
+	 */
+	private $operation;
```
Comment (weierophinney) [inline src/Action.php]: 4 spaces instead of tabs (for PSR1/2). Again, I can fix these during merge.

File: src/Action.php [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+	 * @return string
+	 */
+	public function getResponseDescription() {
+		return $this->operation->getResponseDescription();
+	}
+
+	/**
+	 * @return boolean
+	 */
+	public function isEntityChanging() {
```
Comment (weierophinney) [inline src/Action.php]: Maybe `allowsChangingEntity()`? Since it's a boolean, a natural phrase that implies a yes/no is nice for such methods:

```php
if ($action->allowsChangingEntity()) {
    ...
}
```

File: src/Api.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (weierophinney) [inline src/Api.php]: 2015

File: src/Api.php [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+class Api
+{
+    /**
+     * @var BaseApi
+     */    
+    private $api;
+
+    /**
+     * @var Blueprint\ResourceGroup[]
```
Comment (weierophinney) [inline src/Api.php]: Should be just `ResourceGroup[]`, as that class is in the same namespace.

File: src/Api.php [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+    /**
+     * @return string
+     */
+    public function getName() {
+        return $this->api->getName();
+    }
+
+    /**
+     * @return Blueprint\ResourceGroup[]
```
Comment (weierophinney) [inline src/Api.php]: Same comment here.

File: src/ApiBlueprintModel.php [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+        return true;
+    }
+
+    /**
+     * @return  string
+     */
+    public function getFormattedApiBlueprint() {
+        $model = new Api($this->variables['documentation']);
+        $this->apiBlueprint = 'FORMAT: ' . self::FORMAT . PHP_EOL;
+        $this->apiBlueprint .= 'HOST: ' . $_SERVER["REQUEST_SCHEME"] . "://" . $_SERVER["HTTP_HOST"] . self::$EMPTY_ROW;
```
Comment (weierophinney) [inline src/ApiBlueprintModel.php]: Don't use superglobals within classes. If you need information from superglobals, inject them via the constructor, or via argument parameters. In this particular case, the information here is present in the request instance, which you could either pass to the method, or pass the components you need as arguments directly.

File: src/ApiBlueprintModel.php [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+    /**
+     * @var string
+     */
+    private static $EMPTY_ROW;
+
+    public function __construct() {
+        parent::__construct();
+
+        // PHP < 5.6 is unable to concatenate constants into new constant
+        self::$EMPTY_ROW = PHP_EOL . PHP_EOL;
```
Comment (weierophinney) [inline src/ApiBlueprintModel.php]: I would argue that, because this is intended for consumption via a browser or automated tooling, we can use `\n` instead of `PHP_EOL` throughout.

Comment (weierophinney) [inline src/ApiBlueprintModel.php]: Also, if we use "\n\n", then `EMPTY_ROW` can be a constant. :)

File: src/ApiBlueprintModel.php [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+        $this->apiBlueprint = 'FORMAT: ' . self::FORMAT . PHP_EOL;
+        $this->apiBlueprint .= 'HOST: ' . $_SERVER["REQUEST_SCHEME"] . "://" . $_SERVER["HTTP_HOST"] . self::$EMPTY_ROW; 
+        $this->apiBlueprint .= '# ' . $model->getName() . PHP_EOL;
+        $this->apiBlueprint .= $this->writeFormattedResourceGroups($model->getResourceGroups());
+
+        return $this->apiBlueprint;
+    }
+
+    /**
+     * @param Blueprint\ResourceGroup[] $resourceGroups
```
Comment (weierophinney) [inline src/ApiBlueprintModel.php]: Should be `ResourceGroup[]`.

File: src/ApiBlueprintModel.php [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+     */
+    private function writeFormattedResourceGroups(array $resourceGroups) {
+        foreach ($resourceGroups as $resourceGroup) {
+            $this->apiBlueprint .= '# Group ' . $resourceGroup->getName() . PHP_EOL;
+            $this->writeFormattedResources($resourceGroup->getResources());
+        }
+    }
+
+    /**
+     * @param Blueprint\Resource[] $resources
```
Comment (weierophinney) [inline src/ApiBlueprintModel.php]: Should be `Resource[]`.

File: src/ApiBlueprintModel.php [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+                    $this->writeBodyProperties($resource->getbodyProperties());
+                }
+                $this->writeUriParameters($resource);
+                $this->writeFormattedActions($resource->getActions());
+            }
+        }
+    }
+
+    /**
+     * @param Blueprint\Action[] $resources
```
Comment (weierophinney) [inline src/ApiBlueprintModel.php]: Should be `Action[]`.

File: src/ApiBlueprintModel.php [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+            if($action->isEntityChanging() && !empty($requestDescription)){
+                $this->apiBlueprint .= '+ Request' . self::$EMPTY_ROW;
+                $this->apiBlueprint .= $this->getFormattedCodeBlock($action->getRequestDescription())  . self::$EMPTY_ROW;
+            }
+            $this->writeFormattedResponses($action);
+        }
+    }
+
+    /**
+     * @param Blueprint\Action $action
```
Comment (weierophinney) [inline src/ApiBlueprintModel.php]: Should be `Action`.

File: src/ApiBlueprintModel.php [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+     */
+    private function writeBodyProperties(array $bodyProperties) {
+        foreach ($bodyProperties as $property) {
+            $this->apiBlueprint .= "+ " . $this->getFormattedProperty($property) . PHP_EOL;
+        }
+        $this->apiBlueprint .= self::$EMPTY_ROW;
+    }
+
+    /**
+     * @var \Blueprint\Resource $resource
```
Comment (weierophinney) [inline src/ApiBlueprintModel.php]: Should be `Resource`.

File: src/ApiBlueprintModel.php [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+            }
+        }
+    }
+
+    /**
+     * @var string $codeBlock
+     * @return string
+     */
+    private function getFormattedCodeBlock($codeBlock) {
+        return self::CODE_BLOCK_INDENT . \str_replace("\n", "\n" . self::CODE_BLOCK_INDENT, $codeBlock);
```
Comment (weierophinney) [inline src/ApiBlueprintModel.php]: No need to prefix global functions with the `\` qualifier.

File: src/ApiBlueprintRenderer.php [hunk 22]
```diff
@@ -0,0 +1,38 @@
+<?php
+/**
+ * @license   http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ * @copyright Copyright (c) 2014 Zend Technologies USA Inc. (http://www.zend.com)
```
Comment (weierophinney) [inline src/ApiBlueprintRenderer.php]: 2015

File: src/ApiBlueprintRenderer.php [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+    	$this->resolver = $resolver;
+    }
+
+    /**
+     * @param  string|ModelInterface   $nameOrModel The script/resource process, or a view model
+     * @param  null|array|\ArrayAccess $values      Values to use during rendering
+     * @return string The script output.
+     */
+    public function render($nameOrModel, $values = null) {
+    	echo $nameOrModel->getFormattedApiBlueprint();
```
Comment (weierophinney) [inline src/ApiBlueprintRenderer.php]: Replace the `echo` with `return`; renderers should return the output, not echo it directly.

File: src/ApiBlueprintViewStrategy.php [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+        }
+
+        $response = $e->getResponse();
+        if (! method_exists($response, 'getHeaders')) {
+            return;
+        }
+
+        $headers = $response->getHeaders();
+        $headers->addHeaderLine('Content-Type', 'text/vnd.apiblueprint+markdown');
+    }
```
Comment (weierophinney) [inline src/ApiBlueprintViewStrategy.php]: Once you update the renderer to return the results of rendering, you can update this to add:

```php
$response->setContent($e->getResult());
```

File: src/Resource.php [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+	 */
+	private $service;
+
+	/**
+	 * @var string
+	 */
+	private $uri;
+
+	/**
+	 * @var Blueprint\Action[]
```
Comment (weierophinney) [inline src/Resource.php]: Should be `Action[]`.

File: src/Resource.php [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+	private $actions = array();
+
+	/**
+	 * @var self::RESOURCE_TYPE_*
+	 */
+	private $resourceType;
+
+	/**
+	 * @param BaseService $service
+	 * @param BaseOperations[] $operations
```
Comment (weierophinney) [inline src/Resource.php]: Should be `BaseOperation[]`.

File: src/Resource.php [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+		$this->operations = $operations;
+		$this->uri = $uri;
+		$this->createActions();
+		$this->resourceType = $resourceType;
+
+		if ($this->getResourceType() == self::RESOURCE_TYPE_COLLECTION && $this->getParameter()) {
+            $this->uri .= "?page={page}";
+        }
+
+    }
```
Comment (weierophinney) [inline src/Resource.php]: Mixture of spaces and tabs above; I'll fix this with phpcs later.

File: src/Resource.php [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+
+    /**
+     * @return \ZF\Apigility\Documentation\Field[]
+     */
+    public function getBodyProperties() {
+    	return $this->service->getFields();
+    }
+
+    /**
+     * @return Blueprint\Action[]
```
Comment (weierophinney) [inline src/Resource.php]: Should be `Action[]`

File: src/ResourceGroup.php [hunk 29]
```diff
@@ -0,0 +1,77 @@
...
+
+class ResourceGroup
+{
+    /**
+     * @var BaseService
+     */
+    protected $service;
+
+    /**
+     * @var Blueprint\Resource[]
```
Comment (weierophinney) [inline src/ResourceGroup.php]: Should be `Resource[]`.

File: test/ApiTest.php [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+namespace ZFTest\Apigility\Documentation\ApiBlueprint;
+
+use PHPUnit_Framework_TestCase as TestCase;
+use ZF\Apigility\Documentation\ApiBlueprint\Api;
+
+class ApiTest extends TestCase
+{
+public function setUp () {
```
Comment (weierophinney) [inline test/ApiTest.php]: Indentation; I can fix during merge.

**Comment by weierophinney** [comment]
@miiila Awesome! The only real actionable item I have for you is the question about `circle.yml`; I can take care of the other issues during merge, and I'll setup phpcs, Travis-CI, etc at that point as well.

Thanks!

**Comment by miiila** [comment]
@weierophinney Thank for your nice and precise review! I addressed all issues (I hope :-)) except those you can easily fix during merge. I also removed circle.yml file (configuration for Circle CI), because setting up CI on Travis will be better, IMHO.

Please let me know if you will find anything else that would require my attention. Ale keep me updated what will happen now :-)
